akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -101,22 +119,27 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
   }
 
   private def atLeastOnceProcessing(
-      source: Source[(Offset, Envelope), NotUsed],
+      source: Source[ProjectionContextImpl[Offset, Envelope], NotUsed],
       afterEnvelopes: Int,
       orAfterDuration: FiniteDuration,
       recoveryStrategy: HandlerRecoveryStrategy): Source[Done, NotUsed] = {
 
-    val atLeastOnceHandlerFlow: Flow[(Offset, Envelope), (Offset, Envelope), NotUsed] =
+    val atLeastOnceHandlerFlow
+        : Flow[ProjectionContextImpl[Offset, Envelope], ProjectionContextImpl[Offset, Envelope], NotUsed] =
       handlerStrategy match {
         case SingleHandlerStrategy(handler) =>
           val handlerRecovery =
             HandlerRecoveryImpl[Offset, Envelope](projectionId, recoveryStrategy, logger, statusObserver)
 
-          Flow[(Offset, Envelope)].mapAsync(parallelism = 1) {
-            case elem @ (offset, envelope) =>
-              handlerRecovery
-                .applyRecovery(envelope, offset, offset, abort.future, () => handler.process(envelope))
-                .map(_ => elem)
+          Flow[ProjectionContextImpl[Offset, Envelope]].mapAsync(parallelism = 1) { context =>
+            handlerRecovery
+              .applyRecovery(
+                context.envelope,
+                context.offset,
+                context.offset,
+                abort.future,
+                timedProcess(context, () => handler.process(context.envelope)))
```

Comment:
@ignasi35 now I don't understand the relation to ProjectionContext. I thought the idea was that the start time would be added to the ProjectionContextImpl (maybe when it's created?), and then that time would be used in afterProcess to measure the duration.

That approach will also work with FlowHandler, which would be difficult otherwise.

Reply:
This PR should have been a DRAFT.

Reply:
(it was a draft when this was #222 but I pressed the wrong buttons when recreating it)

Reply:
ok, maybe better that you ping me when you want feedback